catkin_ws/src/controls/src/joystick.py

```diff
@@ -1,85 +1,472 @@
+            )
+        if keyboard.is_pressed("i"): 
+            target_quaternion *= np.quaternion(
+                self.pid_quaternion_increment_w,
+                0.0,
+                self.pid_quaternion_increment_xyz,
+                0.0
+            )
+        if keyboard.is_pressed("k"): 
+            target_quaternion *= np.quaternion(
+                self.pid_quaternion_increment_w,
+                0.0,
+                -self.pid_quaternion_increment_xyz,
+                0.0
+            )
+        if keyboard.is_pressed("j"): 
+            target_quaternion *= np.quaternion(
+                self.pid_quaternion_increment_w,
+                0.0,
+                0.0,
+                self.pid_quaternion_increment_xyz
+            )
+        if keyboard.is_pressed("l"): 
+            target_quaternion *= np.quaternion(
+                self.pid_quaternion_increment_w,
+                0.0,
+                0.0,
+                -self.pid_quaternion_increment_xyz
+            )
+
+        if self.mode_current == self.mode_pid_local:
+            # Convert local desired postion (target_[xyz]) to
+            # NWU positon.
+            target_x, target_y, target_z = local_to_global(
+                self.quaternion,
+                [target_x, target_y, target_z]
+            )  
+            # Convert local desired rotation (target_quaternion)  
+            # to NWU rotation (self.quaternion).
+            target_quaternion = self.quaternion * target_quaternion
+        
+        self.pub_pid_x.publish(Float64(target_x))
+        self.pub_pid_y.publish(Float64(target_y))
+        self.pub_pid_z.publish(Float64(target_z))
+        self.pub_pid_quat.publish(
+            Quaternion(
+                target_quaternion.x, 
+                target_quaternion.y, 
+                target_quaternion.z,
+                target_quaternion.w
+            )
+        )
+        target_roll, target_pitch, target_yaw = transformations.euler_from_quaternion(
+            [
+                target_quaternion.w, 
+                target_quaternion.x, 
+                target_quaternion.y, 
+                target_quaternion.z
+            ]
+        ) 
+        pretty_print_pid([target_x, target_y, target_z, target_roll, target_pitch, target_yaw])
+
+    def execute(self):
+        self.establish_key_hooks()
+        print_key_options_msg(self.modes)
+
+        while not (rospy.is_shutdown() or self.is_esc_pressed):
+            if self.wait_execute:
+                rospy.sleep(2)
+            current_time = rospy.get_time()
+            if (
+                any(x is None for x in [self.x, self.y, self.z, self.quaternion]) or 
+                current_time - self.last_pose_received > self.missing_pose_interval
+            ):
+                if current_time - self.last_warning_time > self.missing_pose_interval:
+                    self.last_warning_time = current_time 
+                    rospy.logwarn("Missing pose values.")
```

Comment:
would be helpful to print which ones

Reply:
not really because state estimation will just send 0s

Reply:
whats the point of this check then? checking last pose received is all we need to do not check if x/y/z/quat are none

Reply:
I mean, I haven't finished writing the code yet. I put this as a way to remember to add some pose-check and give a warning since I use PIDs. But this branch is still not ready to be a PR